Add address sanitizer nightly job

bhatsach opened the pull request:
Address sanitizer is a good way to detect address related issues in c/c++
code as mentioned in https://github.com/google/sanitizers/wiki/AddressSanitizer.
This change adds a nightly job to run address sanitizer on rmw based packages.


### Testing done
* Tested the commands locally on a fresh ec2 linux instance
* Created a new job on a cloned CI : http://ec2-18-144-12-181.us-west-1.compute.amazonaws.com/job/ci_linux/84 . Please ping me directly to get the credentials to the buildfarm
* Running a new job with the latest changes: http://ec2-18-144-12-181.us-west-1.compute.amazonaws.com/job/nightly_linux_address_sanitizer

thomas-moulard commented on job_templates/ci_job.xml.em:
```diff
@@ -168,6 +172,12 @@ fi
 @[  if os_name in ['linux', 'linux-aarch64'] and turtlebot_demo]@
 export CI_ARGS="$CI_ARGS --ros1-path /opt/ros/$CI_ROS1_DISTRO"
 @[  end if]@
+if [ "${CI_ENABLE_SANITIZER_TYPE}" != "none" ]; then
```
Nit: I would use an empty string instead of none to avoid having issues if by any chance we end up having CI_ENABLE_SANITIZER_TYPE not being set.

bhatsach replied:
I was unable to identify a good way to translate jenkins build parameter value from 'none' to ''. Given that there is currently no way to set this parameter to '' through UI, let's continue with the above logic

thomas-moulard commented on job_templates/snippet/property_parameter-definition.xml.em:
```diff
@@ -42,5 +42,27 @@ This tests the robustness to whitespace being within the different paths.</descr
+          <description>This can be used to enable sanitizer based builds and tests</description>
+          <choices class="java.util.Arrays$ArrayList">
+            <a class="string-array">
+              <string>@enable_sanitizer_type_default</string>
```
not using a macro would be more clear IMHO
Here, you put everything in the vector, just to remove one, just to print a single value => https://en.wikipedia.org/wiki/YAGNI

bhatsach replied:
Makes sense, I have simplified it a bit by removing the for loop. Will add it back in the PR for tsan job

bhatsach replied:
Makes sense, I have simplified it a bit by avoiding the for loop

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    return cmake_test_pkgs + export_library_preload_for_sanitizer
+
+def get_suffix_test_cmd_based_on_sanitizer_args(args):
+    reset_library_preload = ['export LD_PRELOAD=;']
```
This needs a few lines of comments to explain why we need this + a FIXME because that's not a good long-term solution

osrf_memory_testing is relying on LD_PRELOAD to instrument libc functions such as e.g. `malloc()`. This conflicts with asan as the sanitizers is doing the same thing. In this case, the run will with fail with the message "<FIXME>" when invoked.

// FIXME(ros2/abcd#1234): stop doing that once we have a CMake argument disabling OSRF memory testing.

bhatsach replied:
Makes sense, done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    print('# BEGIN SUBSECTION: build with sanitizer')
+    sanitizer_type = args.enable_sanitizer_type
+    sanitizer_cmake_args = [
+    '-DCMAKE_CXX_FLAGS="-fsanitize=' + sanitizer_type + ' -g"',
```
Maybe having some kind of shared flags between all jobs would be good. Not sure if necessary here.

bhatsach replied:
Used a single sanitizer flags variable for all 4 args. it might be tricky to have it across jobs though

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
     # return 0 if ret_test == 0 and ret_testr == 0 else 1
     return 0
 
+def build_with_sanitizer(cmd, args, job):
```
Could we add some doc to this method? What kind of shell command do we usually expect from this?

bhatsach replied:
Done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    print('# BEGIN SUBSECTION: build with sanitizer')
+    sanitizer_type = args.enable_sanitizer_type
+    sanitizer_cmake_args = [
+    '-DCMAKE_CXX_FLAGS="-fsanitize=' + sanitizer_type + ' -g"',
```
Rule of thumb is if you end up having more than one + in a string expression, format is probably what you should use instead: https://www.geeksforgeeks.org/python-format-function/

"-DCMAKE_CXX_FLAGS='{} -g'".format(sanitizer_type)

It also helps you declaring those strings as top-level constants.

bhatsach replied:
Makes sense, done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    '-DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=' + sanitizer_type + '"'
+    ]
+    index = cmd.index('--cmake-args')
+    cmd[index + 1:index + 1] = sanitizer_cmake_args
```
Not sure I understand what's going there? Is cmd a string? Are we assigning an array to a string and relying on some kind of implicit conversion to string there? Maybe extracting a function and adding a few line of docs would be good to help the reader?

bhatsach replied:
cmd here is actually a list of strings which are then concatenated and run as a single command. I have extracted this logic into a separate function and added doc

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    suffix_test_cmd = get_suffix_test_cmd_based_on_sanitizer_args(args)
+    return prefix_test_cmd + test_cmd + suffix_test_cmd
+
+def get_prefix_test_cmd_based_on_sanitizer_args(args):
```
Some doc would be helpful.

bhatsach replied:
Done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    print('# END SUBSECTION')
+    return ret_build
+
+def update_test_cmd_based_on_sanitizer_args(test_cmd, args):
```
Some doc would be helpful.

bhatsach replied:
Added it

bhatsach replied:
Done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    export_library_preload_for_sanitizer = []
+    if args.enable_sanitizer_type == 'address':
+        # Address Sanitizer runtime needs to be the first in the initial loaded library list
+        export_library_preload_for_sanitizer = ['export LD_PRELOAD=$(ls /usr/lib/x86_64-linux-gnu/libasan.so* | head -1);']
```
Let's use /usr/lib/x86_64-linux-gnu/libasan.so and drop the | head -n 1 - we can do exact matching here and it will less fragile (or not less fragile than randomly selecting the first versioned shared library which may or may not match what's actually got linked with the binary).

bhatsach replied:
This is a bit tricky as there is no generic libasan.so. I can only see libasan.so.4 and libasan.so.4.0.0. I will try to see if there is any other way to identify the path. One naive way to do it might be to run ld on a ros2 generated binary and grepping for libasan, not sure how to do it though

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+    pkgs_regex_arg = ''
+    if args.restrict_san_pkgs_regex:
+        pkgs_regex_arg = '--packages-select-regex ' + args.restrict_san_pkgs_regex
+    test_pkgs = args.colcon_script + ' list ' + pkgs_regex_arg +  '| awk \'{printf $1" "}\''
```
let's use .format() and document what we expect this string to output.

bhatsach replied:
Done

thomas-moulard commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
+        pkgs_regex_arg = '--packages-select-regex ' + args.restrict_san_pkgs_regex
+    test_pkgs = args.colcon_script + ' list ' + pkgs_regex_arg +  '| awk \'{printf $1" "}\''
+    # filter test packages to just cmake based packages so that it is sanitizer comatible
+    cmake_test_pkgs = ['cmake_test_pkgs="$(' + args.colcon_script +
```
This is really dense, what about splitting that in multiple steps, one for each subshell and adding comments?

bhatsach replied:
Done

bhatsach commented on create_jenkins_job.py:
```diff
@@ -193,6 +195,17 @@ def create_job(os_name, job_name, template_file, additional_dict):
 
+        # configure nightly job for testing rmw based packages with address sanitizer on linux
+        if os_name == 'linux':
+            create_job(os_name, 'nightly_' + os_name + '_address_sanitizer', 'ci_job.xml.em', {
```
Moved it to a separate method

bhatsach commented on create_jenkins_job.py:
```diff
@@ -193,6 +195,17 @@ def create_job(os_name, job_name, template_file, additional_dict):
             })
 
+        # configure nightly job for testing rmw based packages with address sanitizer on linux
+        if os_name == 'linux':
```
Moved it to a separate method

bhatsach commented:
@tfoote, can you please review the code and let me know of any comments. Thomas has already approved the PR.
Thanks

tfoote commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +366,89 @@ def build_and_test(args, job):
+
+    # Identify the sanitized packages
+    pkgs_regex_arg = '--packages-select-regex {}'.format(args.restrict_san_pkgs_regex) if args.restrict_san_pkgs_regex else ''
+    sanitized_pkgs = '$({} list {}'.format(args.colcon_script, pkgs_regex_arg) + '| awk \'{printf $1" "}\')'
```
It would be more maintainable if this was done using list comprehension in python instead of going through a bash pipe to awk.

tfoote commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +366,89 @@ def build_and_test(args, job):
+
+    # Identify all cmake packages which depend on/are sanitized packages
+    pkgs_above_sanitized_pkgs = '{} list --packages-above {}'.format(args.colcon_script, sanitized_pkgs)
+    cmake_pkgs_filter = 'awk \'$3 ~ /cmake/ {printf $1" "}\''
```
It would be more maintainable and readable if this was done using list comprehension in python instead of going through a bash pipe to awk.

allenh1 commented on create_jenkins_job.py:
```diff
@@ -127,6 +129,25 @@ def create_job(os_name, job_name, template_file, additional_dict):
 
+    def create_linux_based_nightly_jobs():
+        # configure nightly job for compiling with clang on linux
+        create_job('linux', 'nightly_linux_clang', 'ci_job.xml.em', {
```
So I think this can be applied to more than just Linux, and, specifically, more than just clang.

Perhaps we could add a GCC address sanitizer build as well? I think MSVC supports it as well, but I'd need to double check that statement.

I really like this idea, by the way. Sanitizers are super useful.

bhatsach replied:
Thanks Allen. Currently I have added a GCC address sanitizer on linux. I will be adding thread sanitizer as well soon. I am not adding any clang sanitizers for now.

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -319,14 +320,23 @@ def build_and_test(args, job):
 
+    if args.enable_sanitizer_type:
+        ret_build_with_sanitizer = build_with_sanitizer(cmd, args, job)
+    if ret_build_with_sanitizer:
```
This line will fail if sanitizer is not enabled. The condition should be indented one level.

bhatsach replied:
Good catch! I have avoided this logic for now

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -319,14 +320,23 @@ def build_and_test(args, job):
         return ret_build
 
+    if args.enable_sanitizer_type:
+        ret_build_with_sanitizer = build_with_sanitizer(cmd, args, job)
```
How much overhead does building the selected packages a second introduce? It would be better to build them with the desired flags in the first pass.

bhatsach replied:
Good point, I was experimenting the same while working on thread sanitizer job. I have modified the build command to perform the build in three steps: build till regex packages, build regex packages with sanitizer, build above regex packages. This avoids the additional pass of colcon build on all packages

bhatsach commented:
> This looks reasonable. There are a few cleanups I flagged. One is the use of bash pipes to awk instead of pythonic list comprehension. The list of valid sanitizers is only available inside the drop down menu and is not validated by the tests themselves. This is expected to work on the farm every time, but it's fragile if someone tries to reproduce locally w/o using the jenkins UI. And I think that the scope on the LD_PRELOAD should be reviewed.

Thanks for the feedback, all the points make sense. I have made the following changes:
1/ Moved package name extraction and filtering logic to python from awk
2/ Moved list of sanitizers to python code in utils class
3/ Instead of exporting LD_PRELOAD, I am just setting it as a variable for the colcon test command

bhatsach commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +366,89 @@ def build_and_test(args, job):
+'''
+def get_test_cmd_suffix_based_on_sanitizer_args(args):
+    select_cmake_pkgs_for_testing = ['--packages-select $cmake_test_pkgs;']
+    reset_library_preload = ['export LD_PRELOAD=;']
```
Good point, changed to just setting LD_PRELOAD as a variable before calling colcon test

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +366,104 @@ def build_and_test(args, job):
+    '''
+    Build command is split into build upto regex packages, build regex packages and then build
+    packages above regex packages. This ensures that we do not rebuild all the packages again with sanitizer flags.
+    '''
```
This approach is not going to work for an arbitrary regex which could select packages throughout the dependency graph.

bhatsach replied:
Can you give an example here? I am sorry I am new to to colcon build concepts :(.

dirk-thomas replied:
If you have three packages with a dependency chain:

`A -> B -> C`

And the regex matches A and C.

bhatsach replied:
Oh got it, thanks. We may be alright to continue with this logic as we ultimately want to test sanitization on just A and C and hence do not need B to be sanitized. Also, in the previous approach(build all without sanitizer flags and build just regex packages with sanitizer flags), we ended up in a similar state where A and C are sanitized but B is not. Let me know if I am missing something here.

It is definitely true that there may be some occasions in which both the approaches might break. For example, if colcon test on sanitized A needs B to be sanitized as well.

dirk-thomas replied:
The approach will still build some of the packages twice. Also hard coding all this extra logic in this repo doesn't look appealing to me.

bhatsach replied:
> Taking a step back I would prefer an approach where a user can perform the same build locally - without having to duplicate all this logic / settings. So e.g. all the necessary CMake arguments could be stored in a colcon mixin file (as it is already the case for some: https://github.com/colcon/colcon-mixin-repository/blob/3e627e0fa30db85aea05a50e2c61a9832664d236/asan.mixin - not sure if the differences are necessary and should be added to the existing mixin?). Based on that there should be a way to specify that the mixin should be applied to a set of packages (needs to be implenmnted) and after that a single build pass should be all it takes.
> 
> This would make it easy for users to run the very same locally as well as reduce the custom logic necessary to integration it into the Jenkins jobs.

Makes sense. This way all the packages will be built only once. I am assuming this would be a change in colcon itself (adding an option to add a mixin for a set of packages). Do you think we should handle it as a separate issue?

bhatsach replied:
> The approach will still build some of the packages twice. Also hard coding all this extra logic in this repo doesn't look appealing to me.

One naive way to avoid the rebuild would be something this:
Step 1:   Build all pkgs upto regex and skip regex pkgs
Step 2:   Build regex pkgs
Step 3: Build pkgs above regex pkgs

Definitely ugly but will avoid the rebuild

dirk-thomas replied:
> I am assuming this would be a change in colcon itself (adding an option to add a mixin for a set of packages). 

Yes, that would require support for that use case in `colcon`. I am just not sure how exactly it would look like. The complexity for controlling this using command line options is too high imo. I will need to think about how this fits in with the already existing feature like metadata, mixins, defaults as well as other feature ideas like profiles...

> One naive way to avoid the rebuild would be something this:

That approach doesn't work which I tried to point out above. If your regex matched packages have non-matches packages in between 3 steps are not sufficient. In a trivial case you would need 5 and in a more complex case up to N separate steps.

bhatsach replied:
What is the downside of not sanitizing the packages in between two regex matched packages? Instrumentation will not be performed on these packages and hence normal malloc() and other functions will be called. Are we targetting the use case where all regex matched packages and the packages between them are sanitized?

dirk-thomas replied:
> Step 1: Build all pkgs upto regex and skip regex pkgs
> Step 2: Build regex pkgs
> Step 3: Build pkgs above regex pkgs

The problem is not that some packages are not sanitized. The build in the first step will just fail. In my previously mentioned example (A -> B -> C where the regex matches A and C) step 1 tries to build A and C. But since B hasn't been built yet but is a dependency of A the build will fail.

bhatsach replied:
Sorry for another rookie question, but I assumed that B will be built before A is built. Is it not what "--packages-up-to" is supposed to do? 
  --packages-up-to [PKG_NAME [PKG_NAME ...]]
                        Only process a subset of packages and their recursive
                        dependencies

dirk-thomas replied:
Your step explicitly states: `Build all pkgs upto regex and skip regex pkgs`.

So when you select "up to" A and C you will select all three packages and applying the "skip regex pkgs" you will only have package B left. But that package can't be built without its dependency C being built first.

bhatsach replied:
Ahh.. thanks a lot for pointing it out. Makes sense, so skipping is not an option. Building the regex packages twice is safer.

bhatsach commented on create_jenkins_job.py:
```diff
@@ -184,14 +205,8 @@ def create_job(os_name, job_name, template_file, additional_dict):
-                'time_trigger_spec': PERIODIC_JOB_SPEC,
-                'mailer_recipients': DEFAULT_MAIL_RECIPIENTS,
-            })
+            create_linux_based_nightly_jobs()
```
Done, moved back to the original function

bhatsach commented on create_jenkins_job.py:
```diff
@@ -301,6 +316,5 @@ def create_job(os_name, job_name, template_file, additional_dict):
     job_config = expand_template('ci_launcher_job.xml.em', job_data)
     configure_job(jenkins, 'ci_launcher', job_config, **jenkins_kwargs)
 
-
```
Done, removed this

dirk-thomas commented:
Taking a step back I would prefer an approach where a user can perform the same build locally - without having to duplicate all this logic / settings. So e.g. all the necessary CMake arguments could be stored in a colcon mixin file (as it is already the case for some: https://github.com/colcon/colcon-mixin-repository/blob/3e627e0fa30db85aea05a50e2c61a9832664d236/asan.mixin - not sure if the differences are necessary and should be added to the existing mixin?). Based on that there should be a way to specify that the mixin should be applied to a set of packages (needs to be implenmnted) and after that a single build pass should be all it takes.

This would make it easy for users to run the very same locally as well as reduce the custom logic necessary to integration it into the Jenkins jobs.

bhatsach commented:
@tfoote, I have incorporated the comment, can you please take a look when you get time?

bhatsach commented:
Thanks @tfoote and @dirk-thomas. Can you please help with a test run and merging of the changes?

dirk-thomas commented:
> Can you please help with a test run and merging of the changes?

Imo this shouldn't be merged as is since the patch is very hacky and I would rather wait for a clean implementation of the feature (simply because the past has shown numerous times that it is unlikely to be followed up on once a first version is there). Anyway I will leave it up to @tfoote to test the current changes and merge them anyway.

> all the necessary CMake arguments could be stored in a colcon mixin file (as it is already the case for some: https://github.com/colcon/colcon-mixin-repository/blob/3e627e0fa30db85aea05a50e2c61a9832664d236/asan.mixin - not sure if the differences are necessary and should be added to the existing mixin?).

@bhatsach Could you please answer my previous question about the need of the arguments in build invocation compared to the mixin.

bhatsach commented:
> @bhatsach Could you please answer my previous question about the need of the arguments in build invocation compared to the mixin.

The flags, '-g' and '-fno-omit-frame-pointer', are needed to get better stack traces and file names in error messages (suggested here: https://clang.llvm.org/docs/AddressSanitizer.html#usage). Without the the '-pthread' flag, we are running into undefined reference errors like the below ones while building packages like fastrtps/fastcdr/tf2_ros with asan:
.... undefined reference to symbol 'pthread_create'...

dirk-thomas commented:
> The flags, '-g' and '-fno-omit-frame-pointer', are needed to get better stack traces and file names in error messages (suggested here: https://clang.llvm.org/docs/AddressSanitizer.html#usage).

Please consider to provide a pull request to add these arguments to the existing mixin.

> Without the the '-pthread' flag, we are running into undefined reference errors like the below ones while building packages like fastrtps/fastcdr/tf2_ros with asan

A similar problem was discussed in colcon/colcon-mixin-repository#12 for `googletest` and `osrf_testing_tools_cpp`. If additional packages need similar patches please consider to create pull requests for them. I don't think that the build invocation should be responsible to account for those.

dirk-thomas commented:
> The flags, '-g' and '-fno-omit-frame-pointer', are needed to get better stack traces and file names in error messages (suggested here: https://clang.llvm.org/docs/AddressSanitizer.html#usage).

Thanks for the answer. Please consider to provide a pull request to add these arguments to the existing mixin.

> Without the the '-pthread' flag, we are running into undefined reference errors like the below ones while building packages like fastrtps/fastcdr/tf2_ros with asan

A similar problem was discussed in colcon/colcon-mixin-repository#12 for `googletest` and `osrf_testing_tools_cpp`. If additional packages need similar patches please consider to create pull requests for them. I don't think that the build invocation should be responsible to account for those.

bhatsach commented:
@dirk-thomas, @tfoote and @thomas-moulard, I would like to propose an alternate approach here so that we do just a single build pass. 
### Approach
* User provides a regex for eg: 'rmw|rcl|test'
* We build all packages upto the above regex packages with sanitization and then test the same
* We do not build any packages above the regex packages

### Reasons for choosing the approach
* In order to get thread sanitizer checks working, we need all the dependencies of a package to be sanitized with tsan as well. Hence, we may anyway end up making above changes for thread sanitizer if not for address sanitizer
* Building and testing all packages above a particular regex may end including almost all the ros2 packages, hence resulting in long running jobs and huge reports which are difficult to analyze and track.

Please let me know your thoughts on this

dirk-thomas commented:
> We do not build any packages above the regex packages

If your regex will match the rmw packages (as described before) that wouldn't be sufficient since there would be no code actually using these packages. Therefore I think you do need to build and test at least some packages above.

One naive question would be: why not build all packages with sanitizer flags enables? Each build can still limit the set of packages to a reasonable subset in order to not build/test all packages.

bhatsach commented:
> > We do not build any packages above the regex packages
> 
> If your regex will match the rmw packages (as described before) that wouldn't be sufficient since there would be no code actually using these packages. Therefore I think you do need to build and test at least some packages above.
> 
> One naive question would be: why not build all packages with sanitizer flags enables? Each build can still limit the set of packages to a reasonable subset in order to not build/test all packages.

I quickly tested with the regex: 'rmw|rcl|test' and it includes the following packages: rcl rcl_action rcl_interfaces rcl_lifecycle rcl_logging_noop rcl_yaml_param_parser rclcpp rclcpp_action rclcpp_lifecycle rclpy rmw rmw_fastrtps_cpp rmw_fastrtps_shared_cpp rmw_implementation rmw_implementation_cmake test_cli test_cli_remapping test_communication test_launch_ros test_msgs test_osrf_testing_tools_cpp test_rclcpp test_security
I could generate a reasonable report of tsan failures with tests on these packages

Yeah, that is the approach I am suggesting. For eg: if the dependency graph is A -> B -> C -> D -> E and the regex matches B and D, then we end up building B, C, D and E with sanitization and running tests on B and D. Is this the approach you mentioned as well?

dirk-thomas commented:
> Is this the approach you mentioned as well?

No. I was wondering why the sanitizer needs to be limited to a subset of packages. It would be much easier to just build all packages (not necessarily all packages in ros2.repos but all packages selected by a build) with the sanitizer options. You can then still select whatever subset of packages to run the tests on.

The point is that the two features (1) package selection what to build as well as (2) package selection what to test are already available. The only step missing would be to configure the sanitizer options globally.

bhatsach commented:
> > Is this the approach you mentioned as well?
> 
> No. I was wondering why the sanitizer needs to be limited to a subset of packages. It would be much easier to just build all packages (not necessarily all packages in ros2.repos but all packages selected by a build) with the sanitizer options. You can then still select whatever subset of packages to run the tests on.
> 
> The point is that the two features (1) package selection what to build as well as (2) package selection what to test are already available. The only step missing would be to configure the sanitizer options globally.

Makes sense, I am in on this approach. I was mainly trying to avoid unnecessary build of packages which we may never test.
Based on the above approach, these could be the changes:
* Use the already available CI_BUILD_ARGS and CI_TEST _ARGS to build and test all/subset of packages
* If sanitizer is enabled -> add necessary flags during build (or use a global approach to fetch the flags), add ld_preload variable before test and limit test to just cmake packages

bhatsach commented:
Tested the above approach for address sanitizer and here's the report: http://ec2-18-144-12-181.us-west-1.compute.amazonaws.com/job/nightly_linux_sanitizer/32/testReport/ . Definitely a reasonable approach :). I will go ahead and update the PR with the changes if everyone is onboard

dirk-thomas commented:
> these could be the changes:
> Use the already available CI_BUILD_ARGS and CI_TEST _ARGS to build and test all/subset of packages

Just to clarify: what kind of changes do you plan to implement here?

> If sanitizer is enabled

Adding a new boolean flag to enable the address sanitizer or an option to select "address sanitizer" (and in the future others) sounds good.

> add necessary flags during build (or use a global approach to fetch the flags)

This should certainly use the mixin.

> add ld_preload variable before test

:+1:

> limit test to just cmake packages

Why does this need to happen? The user can simply pass the desired set of package names to be tested.

bhatsach commented:
> > these could be the changes:
> > Use the already available CI_BUILD_ARGS and CI_TEST _ARGS to build and test all/subset of packages
> 
> Just to clarify: what kind of changes do you plan to implement here?

To start off with, let's not modify the CI_BUILD_ARGS so that all packages will be built. In the future, we can use options like "packages-up-to test_communication" in order to build till particular package(s)

> 
> > If sanitizer is enabled
> 
> Adding a new boolean flag to enable the address sanitizer or an option to select "address sanitizer" (and in the future others) sounds good.
> 
> > add necessary flags during build (or use a global approach to fetch the flags)
> 
> This should certainly use the mixin.
Will talk to @thomas-moulard about this offline as I don't know how to use the mixins. Is there any good readme/doc which can help with the usage of mixin's during colcon build?

> 
> > add ld_preload variable before test
> 
> 👍
> 
> > limit test to just cmake packages
> 
> Why does this need to happen? The user can simply pass the desired set of package names to be tested.
Yeah, agreed. We can skip this if we mention all the required target test packages using "--packages-select" in CI_TEST_ARGS. Just that if anyone uses the "--packages-select-regex" feature here and python packages are picked up for testing, then the job will fail.

bhatsach commented:
Updated the PR with the above changes. Things look way more simpler and cleaner now :)

dirk-thomas commented on job_templates/snippet/property_parameter-definition.xml.em:
```diff
@@ -42,5 +42,18 @@ This tests the robustness to whitespace being within the different paths.</descr
+@{
+sanitizer_types.remove(enable_sanitizer_type_default)
+}@
+              <string>@sanitizer_types[0]</string>
```
This single line should be replaced with a loop over the remaining `sanitizer_types`.

bhatsach replied:
Removed all sanitizer specific logic now as it is not required anymore.

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -524,6 +564,10 @@ def run(args, build_function, blacklisted_package_names=None):
         print('# END SUBSECTION')
 
+        # Fetch colcon mixins
+        job.run([args.colcon_script, 'mixin add default ' + MIXIN_REPO_URL], shell=True)
```
Each command line argument should be a separate item in the list: `[args.colcon_script, 'mixin', 'add', 'default', MIXIN_REPO_URL]`

Same on the next line.

bhatsach replied:
Made the change. Just for learning sake, why do we want to separate each word into an element of its own?

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +370,32 @@ def build_and_test(args, job):
 
+'''
+Adds sanitizer specific mixin as a build arg
+'''
```
The docblock of a function needs to be in the function block and should use triple double quotes.

Same below.

bhatsach replied:
Removed all sanitizer specific logic from code

dirk-thomas commented on ros2_batch_job/__main__.py:
```diff
@@ -356,6 +370,32 @@ def build_and_test(args, job):
+    application or manually preload it with LD_PRELOAD"
+
+    We are overriding LD_PRELOAD with the sanitizer library path before colcon test to resolve this.
+    FIXME(osrf/osrf_testing_tools_cpp#23): Add a cmake argument for disabling memory tool tests and
```
Please consider to provide a pull request to add the described CMake option and avoid this.

bhatsach replied:
Opened 3 pull requests to add the required cmake option:
https://github.com/ros2/rcutils/pull/135
https://github.com/osrf/osrf_testing_tools_cpp/pull/24
https://github.com/ros2/system_tests/pull/331

We do not need any such override logic in the code anymore

dirk-thomas commented on create_jenkins_job.py:
```diff
@@ -186,13 +192,24 @@ def create_job(os_name, job_name, template_file, additional_dict):
+                'enable_sanitizer_type_default': 'address',
+                'time_trigger_spec': PERIODIC_JOB_SPEC,
+                'mailer_recipients': DEFAULT_MAIL_RECIPIENTS,
+                'test_args_default': '--event-handlers console_direct+ --executor sequential --packages-up-to ' + ' '.join(RMW_RCL_TEST_PKGS),
```
Instead of copy-n-pasting the list of arguments please use the existing variable and only list the additional arguments here.

Same next line.

bhatsach replied:
I am avoiding the "retest until pass" arg from the default set of args as this might unnecessarily slow down asan/tsan test runs

dirk-thomas commented on create_jenkins_job.py:
```diff
@@ -186,13 +192,24 @@ def create_job(os_name, job_name, template_file, additional_dict):
         # configure nightly job for compiling with clang on linux
         if os_name == 'linux':
-            create_job(os_name, 'nightly_' + os_name + '_clang', 'ci_job.xml.em', {
+            create_job('linux', 'nightly_linux_clang', 'ci_job.xml.em', {
```
There is no need to change this.

The next block should also use the variable `os_name` as part of the job name. That will allow enabling the same job on other platforms in the future without having to touch this again.

bhatsach replied:
Changed

bhatsach commented:
@dirk-thomas, thanks for the quick review. I have opened the relevant pull requests in the other packages and simplified the code here. Here's a test run with the asan mixin which passed successfully: 
http://ec2-18-144-12-181.us-west-1.compute.amazonaws.com/job/nightly_linux_sanitizer_mixin/2/

bhatsach commented:
List of pull requests this change is depending on:
Fast-RTPS: https://github.com/eProsima/Fast-RTPS/pull/431
RCUtils: https://github.com/ros2/rcutils/pull/135
OSRF_Testing_Tools: https://github.com/osrf/osrf_testing_tools_cpp/pull/24
ROS2 System Tests: https://github.com/ros2/system_tests/pull/331
Colcon Mixin: https://github.com/colcon/colcon-mixin-repository/pull/14 (not a hard dependency as this PR only adds flags to output nicer stack trace messages)

bhatsach commented on create_jenkins_job.py:
```diff
@@ -40,6 +40,9 @@
 DEFAULT_REPOS_URL = 'https://raw.githubusercontent.com/ros2/ros2/master/ros2.repos'
 DEFAULT_MAIL_RECIPIENTS = 'ros2-buildfarm@googlegroups.com'
 PERIODIC_JOB_SPEC = '30 7 * * *'
+RMW_RCL_TEST_PKGS = ['rcl','rcl_action','rcl_interfaces','rcl_lifecycle','rcl_logging_noop','rcl_yaml_param_parser','rclcpp','rclcpp_action','rclcpp_lifecycle','rclpy',
```
This list needs to be revisited. @nuclearsandwich to help out here

bhatsach commented:
I have closed the following PRs which tried to skip memory tool tests:
RCUtils: [ros2/rcutils#135](https://github.com/ros2/rcutils/pull/135)
OSRF_Testing_Tools: [osrf/osrf_testing_tools_cpp#24](https://github.com/osrf/osrf_testing_tools_cpp/pull/24)

We have a new PR to enable address sanitizer to work with memory tools:
https://github.com/osrf/osrf_testing_tools_cpp/pull/25

nuclearsandwich commented:
FYI I'm working on a local copy of this branch to

* Move more cmake arguments into the asan-gcc mixin
* Prune down the package list
* Make the mixin index URL a Jenkins parameter.

nuclearsandwich commented:
I've opened #250 which we'll likely want to rebase this PR onto once it lands.

bhatsach commented:
> I've opened #250 which we'll likely want to rebase this PR onto once it lands.

Awesome, thanks for helping out @nuclearsandwich :). Let me know if we can close out on the following PRs:
* https://github.com/colcon/colcon-mixin-repository/pull/14
* https://github.com/colcon/colcon-mixin-repository/pull/15
* https://github.com/osrf/osrf_testing_tools_cpp/pull/25

Once that is done and the [Fast-RTPS FindThreads PR](https://github.com/eProsima/Fast-RTPS/pull/438) is merged, we should be unblocked on this PR :)

bhatsach commented:
Merged in [#250](https://github.com/ros2/ci/pull/250) changes

nuclearsandwich commented:
I think we're getting really close here. I pruned down the package to the minimal necessary packages for `--packages-up-to` and then pulled testing back to just test_communication and test_rclcpp although we'll probably add back a few more.

@bhatsach I think we should move the last custom cmake arguments into the mixin since we want those to be used when building with asan locally too.

After that and the LD_PRELOAD tomfoolery needed to be good neighbors with osrf_memory_tools_testing we should be ready to start running this job.

bhatsach commented:
> I think we're getting really close here. I pruned down the package to the minimal necessary packages for `--packages-up-to` and then pulled testing back to just test_communication and test_rclcpp although we'll probably add back a few more.
> 
> @bhatsach I think we should move the last custom cmake arguments into the mixin since we want those to be used when building with asan locally too.
> 
> After that and the LD_PRELOAD tomfoolery needed to be good neighbors with osrf_memory_tools_testing we should be ready to start running this job.

Awesome!! Thanks a lot for the immense help @nuclearsandwich . I have moved the asan specific cmake flag to asan mixin and updated the PR here.

nuclearsandwich commented:
@bhatsach to see where we were at I ran a job yesterday afternoon using our Fast-RTPS branch nut leaving the ROS 2 projects at their current master (so the LD_PRELOAD shim in osrf_memory_testing_tools wasn't used).

The results of that test run are here https://ci.ros2.org/job/ci_linux/6300/ I'm less concerned about the volume of test failures and more curious what form the asan reports take in this build. When a build is running successfully what form will those reports take? Are they going to be easily extractable as a text artifact or do they only appear in stdout or stderr and need to be extracted by a Warnings plugin extension or similar?

bhatsach commented:
> The results of that test run are here https://ci.ros2.org/job/ci_linux/6300/ I'm less concerned about the volume of test failures and more curious what form the asan reports take in this build. When a build is running successfully what form will those reports take? Are they going to be easily extractable as a text artifact or do they only appear in stdout or stderr and need to be extracted by a Warnings plugin extension or similar?

Thanks for triggering the test job run. I can see the ASan/Tsan logs at the following location during my custom buildfarm runs: ***~/workspace/nightly_linux_address_sanitizer_latest/ws/log/latest_test/test_communicatiom/stdout.log***. I am not fully sure on how the warnings plugin works but we will basically need a mechanism to parse the above log and generate appropriate asan reports. We can have this report generation logic in the ros2 ci job code wherein it looks into the above location for logs, parses the relevant content and generates a new report file. Writing a plugin for the same might be better for the long term.

nuclearsandwich commented:
> I am not fully sure on how the warnings plugin works but we will basically need a mechanism to parse the above log and generate appropriate asan reports.

It's basically a groovy script that runs and does regular expression based processing of build output. Here's our configuration for processing CMake warnings https://github.com/ros-infrastructure/buildfarm_deployment/blob/eb797fcc79317b298b474422c73960a542fe7f8a/modules/jenkins_files/files/var/lib/jenkins/hudson.plugins.warnings.WarningsPublisher.xml

> We can have this report generation logic in the ros2 ci job code wherein it looks into the above location for logs, parses the relevant content and generates a new report file. Writing a plugin for the same might be better for the long term.

Since @thomas-moulard and co are going to be the primary consumers of this job, I'd lean on his input to determine whether we need the parser integrated into the job before we start running it.

thomas-moulard commented:
@nuclearsandwich 
A few thoughts regarding the reporting:
- let's not block this on having a proper report, we can always download raw logs and grep, it's "good enough" to get us started
- valgrind and asan/tsan output is very similar. There is no Jenkins plugin AFAIK for TSAN/ASAN but maybe we could recycle some Valgrind plugin for this.
- in an ideal world, I'd rather have a CLI tool so that I can analyze data locally but I understand there would be an overhead developing that and then wrapping it somehow with the Jenkins infra.

What do you think?

nuclearsandwich commented:
>     * in an ideal world, I'd rather have a CLI tool so that I can analyze data locally but I understand there would be an overhead developing that and then wrapping it somehow with the Jenkins infra.

There would indeed be overhead, but I think that whoever contributes the tooling has a heavyweight say in how it's structured.

The three ways I see for incorporating the Sanitizer jobs into the Jenkins job reports

1. A custom parser for the Warnings plugin as described in https://github.com/ros2/ci/pull/245#issuecomment-468775443
This is probably the lowest code solution although it would be difficult to structure in a way that could be utilized for builds outside of Jenkins.

2. A script to process report output and generate a junit xml report. Our existing test infrastructure generates JUnit-formatted XML output which is processed by the Jenkins xUnit plugin. Generating this would probably be the cleanest approach for integrating *San errors as test failures on Jenkins. One challenge I foresee here is that the Sanitizers don't necessarily run tests of their own, and rather run sidecar to our existing tests, which means there's no easy per-package / test unit method to process reports from build output. I have been day-dreaming of a way to give the sanitizers test meta packages which run the tests from selected other packages using the sanitizer runtime, but I've not investigated that approach for feasibility. It's also possible to run a script at the end of a successful build which parses all the build logs from colcon and generates new JUnit XML from each Sanitizer report.

3. Write a (or find) a suitable Jenkins plugin for processing the output from a sanitizer run. There is a [valgrind plugin for Jenkins](https://plugins.jenkins.io/valgrind) that I have absolutely no experience with but it claims to process valgrind XML reports. If that valgrind plugin doesn't magically work for our case then this is probably the highest effort path forward.

>   let's not block this on having a proper report, we can always download raw logs and grep, it's "good enough" to get us started

I agree with this iterative approach. Although it's worth mentioning that I think it would be unreasonable to ask contributors to make sure that their contributions conform to sanitizer output if we don't have infrastructure organizing that output. So something like the approaches mentioned above definitely ought to go on the roadmap in order for these to produce a job we can point all contributors to and "enforce".

dirk-thomas commented:
As mentioned before for actually running the build locally showing the results in Jenkins is not the only goal. At least as important is to be able to make these results "visible" in a local build.

Is it correct that the sanitizer outputs its information to `stdout` / `stderr`? If yes, it should be trivial for a custom colcon event handler extension to find that output (assuming it is distinguishable somehow) and do whatever we want with it. E.g. write a xUnit compliant result file which will be picked up by `colcon test-result` as well as the Jenkins xUnit plugin.

thomas-moulard commented:
@dirk-thomas That approach SGTM.

Can we split the problem into getting the job turned on (this PR) and then implement the reporting tools as a follow-up action item?

dirk-thomas commented:
> Can we split the problem into getting the job turned on (this PR) and then implement the reporting tools as a follow-up action item?

I am ok with that as long as we have the common understanding to follow up with the second part in a timely manner.

dirk-thomas commented on create_jenkins_job.py:
```diff
@@ -194,6 +194,16 @@ def create_job(os_name, job_name, template_file, additional_dict):
+                'time_trigger_spec': PERIODIC_JOB_SPEC,
+                'mailer_recipients': DEFAULT_MAIL_RECIPIENTS,
+                'build_args_default': data['build_args_default'] + ' --mixin asan --packages-up-to test_communication test_rclcpp',
+                'test_args_default': '--event-handlers console_direct+ --executor sequential --packages-select test_communication test_rclcpp',
```
Maybe I missed it in the discussion but why should the testing be limited to these two packages?

thomas-moulard replied:
My understanding was that @bhatsach was targeting a limited set of packages to avoid `osrf_memory_testing` related issues with the goal of expending support down the road once a technical solution is found for those packages.

bhatsach replied:
@nuclearsandwich helped prune the packages to be tested as seen in the commit: https://github.com/ros2/ci/pull/245/commits/0f502cc7a72ff6284aab59c7f7d7ca359614c3ea. Our intention was to avoid unnecessary packages so that build and test both are reduced to a reasonable size and time.

dirk-thomas replied:
I am not sure what labels all other packages as unnecessary but I get the point of reducing the runtime of the job - it just seems like an arbitrary trade off. Anyway I don't mind.

nuclearsandwich replied:
> I am not sure what labels all other packages as unnecessary

Unnecessary is perhaps too strong a word, but yes the intent is to limit runtime and focus on the communications packages. Where @bhatsach's approach was to grab every package likely to be part of the communications layer, I started with few test packages and plan to grow out.

I'm thinking about moving the package selection to another mixin so its easier to edit it and reproduce the build locally.

nuclearsandwich commented:
Build of test_communication with the asan-gcc mixin and `-DOSRF_TESTING_TOOLS_DISABLE_MEMORY_TOOLS=ON`

[![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6499)](https://ci.ros2.org/job/ci_linux/6499/)

nuclearsandwich commented:
Closing in favor of #267. @bhatsach is preserved there as a co-author.
